{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5f98acec",
   "metadata": {},
   "source": [
    "### This implements a lot of the example here:\n",
    "https://github.com/openai/openai-cookbook/blob/57024c70cff473fb520105e9aea3ab4e514be0df/examples/fine-tuned_qa/olympics-1-collect-data.ipynb\n",
    "\n",
    "\n",
    "### The purpose of this notebook is to build a dataset of text that we can use elsewhere to demonstrate question answering using embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c769d5b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import wikipedia\n",
    "\n",
    "# For wikipedia API documentation see here: https://wikipedia.readthedocs.io/en/latest/code.html\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9abb3a06",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: transformers in /home/tim/anaconda3/lib/python3.8/site-packages (4.26.0)\n",
      "Requirement already satisfied: requests in /home/tim/anaconda3/lib/python3.8/site-packages (from transformers) (2.28.2)\n",
      "Requirement already satisfied: tqdm>=4.27 in /home/tim/anaconda3/lib/python3.8/site-packages (from transformers) (4.59.0)\n",
      "Requirement already satisfied: numpy>=1.17 in /home/tim/anaconda3/lib/python3.8/site-packages (from transformers) (1.20.1)\n",
      "Requirement already satisfied: tokenizers!=0.11.3,<0.14,>=0.11.1 in /home/tim/anaconda3/lib/python3.8/site-packages (from transformers) (0.13.2)\n",
      "Requirement already satisfied: regex!=2019.12.17 in /home/tim/anaconda3/lib/python3.8/site-packages (from transformers) (2022.10.31)\n",
      "Requirement already satisfied: filelock in /home/tim/anaconda3/lib/python3.8/site-packages (from transformers) (3.0.12)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/tim/anaconda3/lib/python3.8/site-packages (from transformers) (20.9)\n",
      "Requirement already satisfied: pyyaml>=5.1 in /home/tim/anaconda3/lib/python3.8/site-packages (from transformers) (5.4.1)\n",
      "Requirement already satisfied: huggingface-hub<1.0,>=0.11.0 in /home/tim/anaconda3/lib/python3.8/site-packages (from transformers) (0.12.0)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in /home/tim/.local/lib/python3.8/site-packages (from huggingface-hub<1.0,>=0.11.0->transformers) (4.4.0)\n",
      "Requirement already satisfied: pyparsing>=2.0.2 in /home/tim/anaconda3/lib/python3.8/site-packages (from packaging>=20.0->transformers) (2.4.7)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/tim/anaconda3/lib/python3.8/site-packages (from requests->transformers) (2.1.1)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/tim/anaconda3/lib/python3.8/site-packages (from requests->transformers) (2.10)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/tim/anaconda3/lib/python3.8/site-packages (from requests->transformers) (1.26.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/tim/.local/lib/python3.8/site-packages (from requests->transformers) (2019.3.9)\n"
     ]
    }
   ],
   "source": [
    "!pip install transformers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e2824aef",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package punkt to /home/tim/nltk_data...\n",
      "[nltk_data]   Package punkt is already up-to-date!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "from typing import Set\n",
    "from transformers import GPT2TokenizerFast\n",
    "\n",
    "import numpy as np\n",
    "from nltk.tokenize import sent_tokenize\n",
    "\n",
    "import nltk\n",
    "nltk.download('punkt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fb0f35e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def get_wiki_page(title):\n",
    "    \"\"\"\n",
    "    Get the wikipedia page given a title\n",
    "    \"\"\"\n",
    "    try:\n",
    "        print(f\"Fetching page '{title}'\")\n",
    "        return wikipedia.page(title)\n",
    "    except wikipedia.exceptions.DisambiguationError as e:\n",
    "        return wikipedia.page(e.options[0])\n",
    "    except wikipedia.exceptions.PageError as e:\n",
    "        return None\n",
    "\n",
    "def recursively_find_all_pages(titles, titles_so_far=set(), title_filter=None):\n",
    "    \"\"\"\n",
    "    Recursively find all the pages that are linked to the Wikipedia titles in the list\n",
    "    \"\"\"\n",
    "    all_pages = []\n",
    "    \n",
    "    titles = list(set(titles) - titles_so_far)\n",
    "    if title_filter:\n",
    "        titles = title_filter(titles)\n",
    "    titles_so_far.update(titles)\n",
    "    for title in titles:\n",
    "        page = get_wiki_page(title)\n",
    "        if page is None:\n",
    "            continue\n",
    "        all_pages.append(page)\n",
    "\n",
    "        new_pages = recursively_find_all_pages(page.links, titles_so_far, title_filter)\n",
    "        for pg in new_pages:\n",
    "            if pg.title not in [p.title for p in all_pages]:\n",
    "                all_pages.append(pg)\n",
    "        titles_so_far.update(page.links)\n",
    "    return all_pages\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2efefefa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a filter to limit pages to genomics topics,\n",
    "# and find all matching wikipedia pages--\n",
    "\n",
    "def filter_genomics_titles(titles):\n",
    "    \"\"\"\n",
    "    Get the titles which are related to genomics, given a list of titles\n",
    "    \"\"\"\n",
    "    titles = [title for title in titles if 'genom' in title.lower()]\n",
    "    \n",
    "    return titles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2f31e920",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching page 'Genomics'\n",
      "Fetching page 'Pharmacogenomics'\n",
      "Fetching page 'Cancer pharmacogenomics'\n",
      "Fetching page 'Pharmacogenomics (journal)'\n",
      "Fetching page 'Personal genomics'\n",
      "Fetching page 'New York Genome Center'\n",
      "Fetching page 'The Cancer Genome Atlas'\n",
      "Fetching page 'International Cancer Genome Consortium'\n",
      "Fetching page 'Chinese Cancer Genome Consortium'\n",
      "Fetching page 'National Center for Cancer Genomics, National Project for Personalized Genomic Medicine, South Korean Ministry of Health and Welfare'\n",
      "Fetching page 'Genomics England'\n",
      "Fetching page 'Genomes'\n",
      "Fetching page 'The Institute for Genomic Research'\n",
      "Fetching page 'Genome Research'\n",
      "Fetching page 'Bacterial genome size'\n",
      "Fetching page 'Metagenome'\n",
      "Fetching page 'Clinical metagenomic sequencing'\n",
      "Fetching page 'Ecogenomics'\n",
      "Fetching page 'Genomic library'\n",
      "Fetching page 'Human genome project'\n",
      "Fetching page 'Genome Reference Consortium'\n",
      "Fetching page 'Wellcome Genome Campus'\n",
      "Fetching page 'Chimpanzee genome project'\n",
      "Fetching page 'Genome Project'\n",
      "Fetching page 'Neanderthal genome project'\n",
      "Fetching page 'Neanderthal Man: In Search of Lost Genomes'\n",
      "Fetching page 'Neanderthal genome'\n",
      "Fetching page 'Whole genome shotgun sequencing'\n",
      "Fetching page 'Celera Genomics'\n",
      "Fetching page 'Genome Therapeutics Corporation'\n",
      "Fetching page 'Integrated Microbial Genomes System'\n",
      "Fetching page 'Genomes OnLine Database'\n",
      "Fetching page 'Genomic Standards Consortium'\n",
      "Fetching page 'Viral metagenomics'\n",
      "Fetching page 'Binning (Metagenomics)'\n",
      "Fetching page 'Fungal genome'\n",
      "Fetching page 'Bacterial genome'\n",
      "Fetching page 'List of sequenced fungi genomes'\n",
      "Fetching page 'List of sequenced protist genomes'\n",
      "Fetching page 'List of sequenced algae genomes'\n",
      "Fetching page 'List of sequenced genomes'\n",
      "Fetching page 'Genome size'\n",
      "Fetching page 'Nuclear genome'\n",
      "Fetching page 'Subgenomic mRNA'\n",
      "Fetching page 'Genomic DNA'\n",
      "Fetching page 'Animal Genome Size Database'\n",
      "Fetching page 'The Evolution of the Genome'\n",
      "Fetching page 'Mitochondrial genome'\n",
      "Fetching page 'UCSC Genome Browser'\n",
      "Fetching page 'UCSC Malaria Genome Browser'\n",
      "Fetching page 'Genome browser'\n",
      "Fetching page 'Ensembl genome database project'\n",
      "Fetching page 'Saccharomyces Genome Database'\n",
      "Fetching page 'Genomic'\n",
      "Fetching page 'Institute of Genomics and Integrative Biology'\n",
      "Fetching page 'Human Genome Organization'\n",
      "Fetching page 'International Mammalian Genome Society'\n",
      "Fetching page 'Mammalian Genome'\n",
      "Fetching page 'Human Genome Organisation'\n",
      "Fetching page 'National Institute of Biomedical Genomics'\n",
      "Fetching page 'Genomic medicine'\n",
      "Fetching page 'International System for Human Cytogenomic Nomenclature'\n",
      "Fetching page 'Genomic imprinting'\n",
      "Fetching page 'Array comparative genomic hybridization'\n",
      "Fetching page 'Comparative genomic hybridization'\n",
      "Fetching page 'Rat Genome Database'\n",
      "Fetching page 'Mouse Genome Database'\n",
      "Fetching page 'National Center for Human Genome Research'\n",
      "Fetching page 'Public Genomics'\n",
      "Fetching page 'Nutritional genomics'\n",
      "Fetching page 'Nutriepigenomics'\n",
      "Fetching page 'Public health genomics'\n",
      "Fetching page 'Public Health Genomics (journal)'\n",
      "Fetching page 'International Human Genome Project'\n",
      "Fetching page '10x Genomics'\n",
      "Fetching page 'Mouse Genome Informatics'\n",
      "Fetching page 'Rat genome database'\n",
      "Fetching page 'Ensembl Genomes'\n",
      "Fetching page 'Triticeae Genomics for Sustainable Agriculture'\n",
      "Fetching page 'Wheat Rust Genomic Improvement'\n",
      "Fetching page 'BMC Genomics'\n",
      "Fetching page 'Philippine Genome Center'\n",
      "Fetching page 'Pan-genome'\n",
      "Fetching page 'Binning (metagenomics)'\n",
      "Fetching page 'List of sequenced archaeal genomes'\n",
      "Fetching page 'Lists of sequenced genomes'\n",
      "Fetching page 'Genome sequence'\n",
      "Fetching page 'Genomic island'\n",
      "Fetching page 'Genome (disambiguation)'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/tim/anaconda3/lib/python3.8/site-packages/wikipedia/wikipedia.py:389: GuessedAtParserWarning: No parser was explicitly specified, so I'm using the best available HTML parser for this system (\"lxml\"). This usually isn't a problem, but if you run this code on another system, or in a different virtual environment, it may use a different parser and behave differently.\n",
      "\n",
      "The code that caused this warning is on line 389 of the file /home/tim/anaconda3/lib/python3.8/site-packages/wikipedia/wikipedia.py. To get rid of this warning, pass the additional argument 'features=\"lxml\"' to the BeautifulSoup constructor.\n",
      "\n",
      "  lis = BeautifulSoup(html).find_all('li')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching page 'Genomic organization'\n",
      "Fetching page 'Genome comparison'\n",
      "Fetching page 'Genome projects'\n",
      "Fetching page 'International Grape Genome Program'\n",
      "Fetching page 'Neanderthal Genome Project'\n",
      "Fetching page 'Bovine Genome Project'\n",
      "Fetching page 'Bovine genome database'\n",
      "Fetching page 'Bovine Genome Sequencing and Analysis Consortium'\n",
      "Fetching page 'Bovine genome'\n",
      "Fetching page '100K Pathogen Genome Project'\n",
      "Fetching page 'Honey Bee Genome Sequencing Consortium'\n",
      "Fetching page 'Honey bee genome'\n",
      "Fetching page 'Sociogenomics'\n",
      "Fetching page 'List of sequenced prokaryotic genomes'\n",
      "Fetching page 'Horse genome'\n",
      "Fetching page 'Dog genome'\n",
      "Fetching page 'History of genomics'\n",
      "Fetching page 'VISTA (comparative genomics)'\n",
      "Fetching page 'Genome evolution'\n",
      "Fetching page 'Genome wide association study'\n",
      "Fetching page 'Genome-wide significance'\n",
      "Fetching page 'Universal Declaration on the Human Genome and Human Rights'\n",
      "Fetching page 'Chimpanzee Genome Project'\n",
      "Fetching page 'Human Genome Project - Write'\n",
      "Fetching page 'Human Genome Project - Read'\n",
      "Fetching page 'Reference genome'\n",
      "Fetching page '1000 Fungal Genomes Project'\n",
      "Fetching page 'Human Pangenome Project'\n",
      "Fetching page 'Africa BioGenome Project'\n",
      "Fetching page 'Earth BioGenome Project'\n",
      "Fetching page '1000 Plant Genomes Project'\n",
      "Fetching page 'Carl R. Woese Institute for Genomic Biology'\n",
      "Fetching page 'List of sequenced eukaryotic genomes'\n",
      "Fetching page 'The Institute for Genome Research'\n",
      "Fetching page 'List of sequenced plant genomes'\n",
      "Fetching page 'List of sequenced animal genomes'\n",
      "Fetching page 'Genome Compiler'\n",
      "Fetching page 'List of sequenced bacterial genomes'\n",
      "Fetching page '100,000 Genomes Project'\n",
      "Fetching page 'NHS Genomic Medicine Centres'\n",
      "Fetching page 'Wellcome Trust Genome Campus'\n",
      "Fetching page 'Estonian Genome Project'\n",
      "Fetching page 'Genome Canada'\n",
      "Fetching page 'BGI Genomics'\n",
      "Fetching page 'Beijing Institute of Genomics'\n",
      "Fetching page 'Chinese National Human Genome Center'\n",
      "Fetching page 'Chinese National Human Genome Center, Shanghai'\n",
      "Fetching page 'Microbial Genome Center'\n",
      "Fetching page 'James D. Watson Institute of Genome Sciences, Zhejiang University'\n",
      "Fetching page 'Cancer Genome Project'\n",
      "Fetching page 'Cancer genome sequencing'\n",
      "Fetching page 'Cancer Genome Atlas'\n",
      "Fetching page 'Genome assembly'\n",
      "Fetching page 'Hybrid genome assembly'\n",
      "Fetching page 'Cancer Genomics Hub'\n",
      "Fetching page 'McDonnell Genome Institute'\n",
      "Fetching page 'Center for Cancer Genomics'\n",
      "Fetching page '$1,000 genome'\n",
      "Fetching page 'Nebula Genomics'\n",
      "Fetching page 'Whole-genome sequencing'\n",
      "Fetching page 'Synthetic genomics'\n",
      "Fetching page 'Synthetic Genomics (company)'\n",
      "Fetching page 'Synthetic genomes'\n",
      "Fetching page 'Synthetic Genomics'\n",
      "Fetching page 'Genome instability'\n",
      "Fetching page 'Epigenomic'\n",
      "Fetching page 'Single cell epigenomics'\n",
      "Fetching page 'Restriction landmark genomic scanning'\n",
      "Fetching page 'Epigenomics (journal)'\n",
      "Fetching page 'Epigenomics AG'\n",
      "Fetching page 'Personal Genomics'\n",
      "Fetching page 'The 1000 Genomes Project'\n",
      "Fetching page 'Human Genome Diversity Project'\n",
      "Fetching page 'Genome-wide association study'\n",
      "Fetching page 'Genome-wide association studies'\n",
      "Fetching page 'Personal Genome Project'\n",
      "Fetching page 'Oncogenomics'\n",
      "Fetching page 'Cancer Genome Anatomy Project'\n",
      "Fetching page 'The Centre for Applied Genomics'\n",
      "Fetching page 'Whole genome amplification'\n",
      "Fetching page 'Whole Genome Amplification'\n",
      "Fetching page 'Personalized Onco-genomics'\n",
      "Fetching page 'Personalized onco-genomics'\n",
      "Fetching page 'Databases for oncogenomic research'\n",
      "Fetching page 'CBio Cancer Genomics Portal'\n",
      "Fetching page 'Roche Cancer Genome Database (RCGDB)'\n",
      "Fetching page 'Complete Genomics'\n",
      "Fetching page 'Full Genome Sequencing'\n",
      "Fetching page 'Full genome sequencing'\n",
      "Fetching page 'Genomic counseling'\n",
      "Fetching page 'Pathway Genomics'\n",
      "Fetching page 'Elective genetic and genomic testing'\n",
      "Fetching page 'Pharmacogenomic'\n",
      "Fetching page 'American College of Medical Genetics and Genomics'\n",
      "Fetching page 'Human Genome'\n",
      "Fetching page 'Human Epigenome Project'\n",
      "Fetching page 'Human epigenome'\n",
      "Fetching page 'International Human Epigenome Consortium'\n",
      "Fetching page 'European Genome-phenome Archive'\n",
      "Fetching page 'Genome Institute of Singapore'\n",
      "Fetching page 'Hong Kong Epigenomics Project'\n",
      "Fetching page 'NCBI Epigenomics'\n",
      "Fetching page 'Epigenome editing'\n",
      "Fetching page 'Epigenome'\n",
      "Fetching page 'Genomics (journal)'\n",
      "Fetching page 'Genome biology'\n",
      "Fetching page 'Genome sequencing'\n",
      "Fetching page 'Metagenomics'\n",
      "Fetching page 'Chemogenomics'\n",
      "Fetching page 'Computational genomics'\n",
      "Fetching page 'Cognitive genomics'\n",
      "Fetching page 'Genome editing'\n",
      "Fetching page 'Off-target effects of genome editing'\n",
      "Fetching page 'Landscape genomics'\n",
      "Fetching page 'Genome project'\n",
      "Fetching page 'Beijing Genomics Institute'\n",
      "Fetching page 'Pathogenomics'\n",
      "Fetching page 'Human genome'\n",
      "Fetching page 'Structural genomics'\n",
      "Fetching page 'Mycobacterium Tuberculosis Structural Genomics Consortium'\n",
      "Fetching page 'Joint Center for Structural Genomics'\n",
      "Fetching page 'Population genomics'\n",
      "Fetching page 'Hologenome theory of evolution'\n",
      "Fetching page 'Genomics of domestication'\n",
      "Fetching page 'Social genomics'\n",
      "Fetching page 'National Human Genome Research Institute'\n",
      "Fetching page 'Genome annotation'\n",
      "Fetching page 'Genomic position'\n",
      "Fetching page 'Genomic database'\n",
      "Fetching page 'Proteogenomics'\n",
      "Fetching page 'Vertebrate and Genome Annotation Project'\n",
      "Fetching page 'Epigenomics'\n",
      "Fetching page 'Biological effects of radiation on the epigenome'\n",
      "Fetching page 'Genome'\n",
      "Fetching page 'Genome engineering'\n",
      "Fetching page 'Genome: The Autobiography of a Species in 23 Chapters'\n",
      "Fetching page 'Genome Biology'\n",
      "Fetching page 'Pangenomics'\n",
      "Fetching page 'DOE Joint Genome Institute'\n",
      "Fetching page 'Comparative genomics'\n",
      "Fetching page 'Nutrigenomics'\n",
      "Fetching page '1000 Genomes Project'\n",
      "Fetching page 'Psychogenomics'\n",
      "Fetching page 'Genome-wide complex trait analysis'\n",
      "Fetching page '1000 Genomes'\n",
      "Fetching page 'Joint Genome Institute'\n",
      "Fetching page 'Functional genomics'\n",
      "Fetching page 'Epistasis and functional genomics'\n",
      "Fetching page 'Human Genome Project'\n",
      "Fetching page 'Toxicogenomics'\n",
      "Fetching page 'Comparative Toxicogenomics Database'\n",
      "Fetching page 'Whole genome sequencing'\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "168"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get all the pages for a topic\n",
    "# THIS TAKES A WHILE\n",
    "pages = recursively_find_all_pages([\"Genomics\"], title_filter=filter_genomics_titles)\n",
    "len(pages)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "330cd190",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://en.wikipedia.org/?oldid=1137582695'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# what is a page?\n",
    "pages[0].revision_id\n",
    "# create a link to the wikipedia page with \n",
    "# https://en.wikipedia.org/?oldid={revision_id}\n",
    "# for example https://en.wikipedia.org/?oldid=1137582695\n",
    "# or url = f\"https://en.wikipedia.org/?oldid={pages[0].revision_id}\"\n",
    "\n",
    "url = f\"https://en.wikipedia.org/?oldid={pages[0].revision_id}\"\n",
    "url"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75a169d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "https://en.wikipedia.org/?oldid=1136190969"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "dbe58ea8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This could be a class that defines which tokenizer to use when an instance is created.\n",
    "\n",
    "tokenizer = GPT2TokenizerFast.from_pretrained(\"gpt2\")\n",
    "\n",
    "def count_tokens(text: str) -> int:\n",
    "    \"\"\"count the number of tokens in a string\"\"\"\n",
    "    return len(tokenizer.encode(text))\n",
    "\n",
    "def reduce_long(\n",
    "    long_text: str, long_text_tokens: bool = False, max_len: int = 590\n",
    ") -> str:\n",
    "    \"\"\"\n",
    "    Reduce a long text to a maximum of `max_len` tokens by potentially cutting at a sentence end\n",
    "    \"\"\"\n",
    "    if not long_text_tokens:\n",
    "        long_text_tokens = count_tokens(long_text)\n",
    "    if long_text_tokens > max_len:\n",
    "        sentences = sent_tokenize(long_text.replace(\"\\n\", \" \"))\n",
    "        ntokens = 0\n",
    "        for i, sentence in enumerate(sentences):\n",
    "            ntokens += 1 + count_tokens(sentence)\n",
    "            if ntokens > max_len:\n",
    "                return \". \".join(sentences[:i][:-1]) + \".\"\n",
    "\n",
    "    return long_text\n",
    "\n",
    "def extract_sections(\n",
    "    wiki_text: str,\n",
    "    title: str,\n",
    "    page_id: int, \n",
    "    max_len: int = 1500,\n",
    "    discard_categories: Set[str] = None,\n",
    ") -> str:\n",
    "    \"\"\"\n",
    "    Extract the sections of a Wikipedia page, discarding the references and other low information sections\n",
    "    \"\"\"\n",
    "    if len(wiki_text) == 0:\n",
    "        return []\n",
    "\n",
    "    # find all headings and the coresponding contents\n",
    "    headings = re.findall(\"==+ .* ==+\", wiki_text)\n",
    "    for heading in headings:\n",
    "        wiki_text = wiki_text.replace(heading, \"==+ !! ==+\")\n",
    "    contents = wiki_text.split(\"==+ !! ==+\")\n",
    "    contents = [c.strip() for c in contents]\n",
    "    assert len(headings) == len(contents) - 1\n",
    "\n",
    "    cont = contents.pop(0).strip()\n",
    "    outputs = [(title, \"Summary\", cont, count_tokens(cont)+4)]\n",
    "    \n",
    "    # discard the discard categories, accounting for a tree structure\n",
    "    max_level = 100\n",
    "    keep_group_level = max_level\n",
    "    remove_group_level = max_level\n",
    "    nheadings, ncontents = [], []\n",
    "    for heading, content in zip(headings, contents):\n",
    "        plain_heading = \" \".join(heading.split(\" \")[1:-1])\n",
    "        num_equals = len(heading.split(\" \")[0])\n",
    "        if num_equals <= keep_group_level:\n",
    "            keep_group_level = max_level\n",
    "\n",
    "        if num_equals > remove_group_level:\n",
    "            if (\n",
    "                num_equals <= keep_group_level\n",
    "            ):\n",
    "                continue\n",
    "        keep_group_level = max_level\n",
    "        if plain_heading in discard_categories:\n",
    "            remove_group_level = num_equals\n",
    "            keep_group_level = max_level\n",
    "            continue\n",
    "        nheadings.append(heading.replace(\"=\", \"\").strip())\n",
    "        ncontents.append(content)\n",
    "        remove_group_level = max_level\n",
    "\n",
    "    # count the tokens of each section\n",
    "    ncontent_ntokens = [\n",
    "        count_tokens(c)\n",
    "        + 3\n",
    "        + count_tokens(\" \".join(h.split(\" \")[1:-1]))\n",
    "        - (1 if len(c) == 0 else 0)\n",
    "        for h, c in zip(nheadings, ncontents)\n",
    "    ]\n",
    "\n",
    "    # Create a tuple of (title, section_name, content, number of tokens)\n",
    "    outputs += [(title, page_id, h, c, t) if t<max_len \n",
    "                else (title, page_id, h, reduce_long(c, max_len), count_tokens(reduce_long(c,max_len))) \n",
    "                    for h, c, t in zip(nheadings, ncontents, ncontent_ntokens)]\n",
    "    \n",
    "    return outputs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "7f14aa74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching page 'Epigenomics'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Token indices sequence length is longer than the specified maximum sequence length for this model (2522 > 1024). Running this sequence through the model will result in indexing errors\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 48 sections in the page Epigenomics\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Summary', 'In biology, epigenetics is the study of stable phenotypic changes (known as marks) that do not involve alterations in the DNA sequence. The Greek prefix epi- (ἐπι- \"over, outside of, around\") in epigenetics implies features that are \"on top of\" or \"in addition to\" the traditional genetic basis for inheritance. Epigenetics most often involves changes that affect the regulation of gene expression, but the term can also be used to describe any heritable phenotypic change. Such effects on cellular and physiological phenotypic traits may result from external or environmental factors, or be part of normal development.\\nThe term also refers to the mechanism of changes: functionally relevant alterations to the genome that do not involve mutation of the nucleotide sequence. Examples of mechanisms that produce such changes are DNA methylation and histone modification, each of which alters how genes are expressed without altering the underlying DNA sequence. Gene expression can be controlled through the action of repressor proteins that attach to silencer regions of the DNA. These epigenetic changes may last through cell divisions for the duration of the cell\\'s life, and may also last for multiple generations, even though they do not involve changes in the underlying DNA sequence of the organism; instead, non-genetic factors cause the organism\\'s genes to behave (or \"express themselves\") differently.One example of an epigenetic change in eukaryotic biology is the process of cellular differentiation. During morphogenesis, totipotent stem cells become the various pluripotent cell lines of the embryo, which in turn become fully differentiated cells. In other words, as a single fertilized egg cell – the zygote – continues to divide, the resulting daughter cells change into all the different cell types in an organism, including neurons, muscle cells, epithelium, endothelium of blood vessels, etc., by activating some genes while inhibiting the expression of others.', 397)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Definitions', 'The term epigenetics in its contemporary usage emerged in the 1990s, but for some years has been used with somewhat variable meanings. A definition of the concept of epigenetic trait as a \"stably heritable phenotype resulting from changes in a chromosome without alterations in the DNA sequence\" was formulated at a Cold Spring Harbor meeting in 2008, although alternate definitions that include non-heritable traits are still being used widely.The term epigenesis has a generic meaning of \"extra growth\" that has been used in English since the 17th century.', 111)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', \"Waddington's canalisation, 1940s\", 'The hypothesis of epigenetic changes affecting the expression of chromosomes was put forth by the Russian biologist Nikolai Koltsov. From the generic meaning, and the associated adjective epigenetic, British embryologist C. H. Waddington coined the term epigenetics in 1942 as pertaining to epigenesis, in parallel to Valentin Haecker\\'s \\'phenogenetics\\' (Phänogenetik). Epigenesis in the context of the biology of that period referred to the differentiation of cells from their initial totipotent state during embryonic development.When Waddington coined the term, the physical nature of genes and their role in heredity was not known. He used it instead as a conceptual model of how genetic components might interact with their surroundings to produce a phenotype; he used the phrase \"epigenetic landscape\" as a metaphor for biological development. Waddington held that cell fates were established during development in a process he called canalisation much as a marble rolls down to the point of lowest local elevation. Waddington suggested visualising increasing irreversibility of cell type differentiation as ridges rising between the valleys where the marbles (analogous to cells) are travelling.In recent times, Waddington\\'s notion of the epigenetic landscape has been rigorously formalized in the context of the systems dynamics state approach to the study of cell-fate. Cell-fate determination is predicted to exhibit certain dynamics, such as attractor-convergence (the attractor can be an equilibrium point, limit cycle or strange attractor) or oscillatory.', 327)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Contemporary', 'Robin Holliday defined in 1990 epigenetics as \"the study of the mechanisms of temporal and spatial control of gene activity during the development of complex organisms.\" Thus, in its broadest sense, epigenetic can be used to describe anything other than DNA sequence that influences the development of an organism.\\nMore recent usage of the word in biology follows stricter definitions. As defined by Arthur Riggs and colleagues, it is \"the study of mitotically and/or meiotically heritable changes in gene function that cannot be explained by changes in DNA sequence.\"The term has also been used, however, to describe processes which have not been demonstrated to be heritable, such as some forms of histone modification. Consequently, there are attempts to redefine \"epigenetics\" in broader terms that would avoid the constraints of requiring heritability. For example, Adrian Bird defined epigenetics as \"the structural adaptation of chromosomal regions so as to register, signal or perpetuate altered activity states.\" This definition would be inclusive of transient modifications associated with DNA repair or cell-cycle phases as well as stable changes maintained across multiple cell generations, but exclude others such as templating of membrane architecture and prions unless they impinge on chromosome function. Such redefinitions however are not universally accepted and are still subject to debate. The NIH \"Roadmap Epigenomics Project\", ongoing as of 2016, uses the following definition: \"For purposes of this program, epigenetics refers to both heritable changes in gene activity and expression (in the progeny of cells or of individuals) and also stable, long-term alterations in the transcriptional potential of a cell that are not necessarily heritable.\" In 2008, a consensus definition of the epigenetic trait, a \"stably heritable phenotype resulting from changes in a chromosome without alterations in the DNA sequence,\" was made at a Cold Spring Harbor meeting.The similarity of the word to \"genetics\" has generated many parallel usages. The \"epigenome\" is a parallel to the word \"genome,\" referring to the overall epigenetic state of a cell, and epigenomics refers to global analyses of epigenetic changes across the entire genome. The phrase \"genetic code\" has also been adapted – the \"epigenetic code\" has been used to describe the set of epigenetic features that create different phenotypes in different cells from the same underlying DNA sequence. Taken to its extreme, the \"epigenetic code\" could represent the total state of the cell, with the position of each molecule accounted for in an epigenomic map, a diagrammatic representation of the gene expression, DNA methylation and histone modification status of a particular genomic region. More typically, the term is used in reference to systematic efforts to measure specific, relevant forms of epigenetic information such as the histone code or DNA methylation patterns.', 577)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Developmental psychology', 'In a sense somewhat unrelated to its use in any biological disciplines, the term \"epigenetic\" has also been used in developmental psychology to describe psychological development as the result of an ongoing, bi-directional interchange between heredity and the environment. Interactive ideas of development have been discussed in various forms and under various names throughout the 19th and 20th centuries. An early version was proposed, among the founding statements in embryology, by Karl Ernst von Baer and popularized by Ernst Haeckel. A radical epigenetic view, known as physiological epigenesis, was developed by Paul Wintrebert. Another variation, probabilistic epigenesis, was presented by Gilbert Gottlieb in 2003. This view encompasses all of the possible developing factors on an organism and how they not only influence the organism and each other but how the organism also influences its own development. Gottlieb gave an example of Rhesus monkeys where infants that did not receive typical maternal care lacked serotonin, which in turn made them more aggressive as they got older. On another note, the long-standing notion \"cells that fire together, wire together\" derives from Hebbian theory which asserts that synaptogenesis, a developmental process with great epigenetic precedence, depends on the activity of the respective synapses within a neural network. Where experience alters the excitability of neurons, increased neural activity has been linked to increased demethylation .The developmental psychologist Erik Erikson wrote of an epigenetic principle in his 1968 book Identity: Youth and Crisis, encompassing the notion that we develop through an unfolding of our personality in predetermined stages, and that our environment and surrounding culture influence how we progress through these stages. This biological unfolding in relation to our socio-cultural settings is done in stages of psychosocial development, where \"progress through each stage is in part determined by our success, or lack of success, in all the previous stages.\"Although empirical studies have yielded discrepant results, epigenetic modifications are thought to be a biological mechanism for transgenerational trauma.', 415)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Molecular basis', \"Epigenetic changes modify the activation of certain genes, but not the genetic code sequence of DNA. The microstructure (not code) of DNA itself or the associated chromatin proteins may be modified, causing activation or silencing. This mechanism enables differentiated cells in a multicellular organism to express only the genes that are necessary for their own activity. Epigenetic changes are preserved when cells divide. Most epigenetic changes only occur within the course of one individual organism's lifetime; however, these epigenetic changes can be transmitted to the organism's offspring through a process called transgenerational epigenetic inheritance.  Moreover, if gene inactivation occurs in a sperm or egg cell that results in fertilization, this epigenetic modification may also be transferred to the next generation.Specific epigenetic processes include paramutation, bookmarking, imprinting, gene silencing, X chromosome inactivation, position effect, DNA methylation reprogramming, transvection, maternal effects, the progress of carcinogenesis, many effects of teratogens, regulation of histone modifications and heterochromatin, and technical limitations affecting parthenogenesis and cloning.\", 232)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'DNA damage', 'DNA damage can also cause epigenetic changes. DNA damage is very frequent, occurring on average about 60,000 times a day per cell of the human body (see DNA damage (naturally occurring)). These damages are largely repaired, however, epigenetic changes can still remain at the site of DNA repair. In particular, a double strand break in DNA can initiate unprogrammed epigenetic gene silencing both by causing DNA methylation as well as by promoting silencing types of histone modifications (chromatin remodeling - see next section). In addition, the enzyme Parp1 (poly(ADP)-ribose polymerase) and its product poly(ADP)-ribose (PAR) accumulate at sites of DNA damage as part of the repair process. This accumulation, in turn, directs recruitment and activation of the chromatin remodeling protein, ALC1, that can cause nucleosome remodeling. Nucleosome remodeling has been found to cause, for instance, epigenetic silencing of DNA repair gene MLH1. DNA damaging chemicals, such as benzene, hydroquinone, styrene, carbon tetrachloride and trichloroethylene, cause considerable hypomethylation of DNA, some through the activation of oxidative stress pathways.Foods are known to alter the epigenetics of rats on different diets. Some food components epigenetically increase the levels of DNA repair enzymes such as MGMT and MLH1 and p53. Other food components can reduce DNA damage, such as soy isoflavones. In one study, markers for oxidative stress, such as modified nucleotides that can result from DNA damage, were decreased by a 3-week diet supplemented with soy. A decrease in oxidative DNA damage was also observed 2 h after consumption of anthocyanin-rich bilberry (Vaccinium myrtillius L.) pomace extract.', 390)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'DNA repair', 'Damage to DNA is very common and is constantly being repaired.  Epigenetic alterations can accompany DNA repair of oxidative damage or double-strand breaks.  In human cells, oxidative DNA damage occurs about 10,000 times a day and DNA double-strand breaks occur about 10 to 50 times a cell cycle in somatic replicating cells (see DNA damage (naturally occurring)).  The selective advantage of DNA repair is to allow the cell to survive in the face of DNA damage.  The selective advantage of epigenetic alterations that occur with DNA repair is not clear.', 119)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Repair of oxidative DNA damage can alter epigenetic markers', \"In the steady state (with endogenous damages occurring and being repaired), there are about 2,400 oxidatively damaged guanines that form 8-oxo-2'-deoxyguanosine (8-OHdG) in the average mamalian cell DNA.  8-OHdG constitutes about 5% of the oxidative damages commonly present in DNA.  The oxidized guanines do not occur randomly among all guanines in DNA. There is a sequence preference for the guanine at a methylated CpG site (a cytosine followed by guanine along its 5' → 3' direction and where the cytosine is methylated (5-mCpG)).  A 5-mCpG site has the lowest ionization potential for guanine oxidation.\\n\\nOxidized guanine has mispairing potential and is mutagenic.  Oxoguanine glycosylase (OGG1) is the primary enzyme responsible for the excision of the oxidized guanine during DNA repair. OGG1 finds and binds to an 8-OHdG within a few seconds.  However, OGG1 does not immediately excise 8-OHdG.  In HeLa cells half maximum removal of 8-OHdG occurs in 30 minutes, and in irradiated mice, the 8-OHdGs induced in the mouse liver are removed with a half-life of 11 minutes.When OGG1 is present at an oxidized guanine within a methylated CpG site it recruits TET1 to the 8-OHdG lesion (see Figure). This allows TET1 to demethylate an adjacent methylated cytosine. Demethylation of cytosine is an epigenetic alteration.  \\nAs an example, when human mammary epithelial cells were treated with H2O2 for six hours, 8-OHdG increased about 3.5-fold in DNA and this caused about 80% demethylation of the 5-methylcytosines in the genome.  Demethylation of CpGs in a gene promoter by TET enzyme activity increases transcription of the gene into messenger RNA.  In cells treated with H2O2, one particular gene was examined, BACE1.  The methylation level of the BACE1 CpG island was reduced (an epigenetic alteration) and this allowed about 6.5 fold increase of expression of BACE1 messenger RNA.\\nWhile six-hour incubation with H2O2 causes considerable demethylation of 5-mCpG sites, shorter times of H2O2 incubation appear to promote other epigenetic alterations. Treatment of cells with H2O2 for 30 minutes causes the mismatch repair protein heterodimer MSH2-MSH6 to recruit DNA methyltransferase 1 (DNMT1) to sites of some kinds of oxidative DNA damage. This could cause increased methylation of cytosines (epigenetic alterations) at these locations.   \\nJiang et al. treated HEK 293 cells with agents causing oxidative DNA damage, (potassium bromate (KBrO3) or potassium chromate (K2CrO4)).  Base excision repair (BER) of oxidative damage occurred with the DNA repair enzyme polymerase beta localizing to oxidized guanines.  Polymerase beta is the main human polymerase in short-patch BER of oxidative DNA damage. Jiang et al. also found that polymerase beta recruited the DNA methyltransferase protein DNMT3b to BER repair sites. They then evaluated the methylation pattern at the single nucleotide level in a small region of DNA including the promoter region and the early transcription region of the BRCA1 gene. Oxidative DNA damage from bromate modulated the DNA methylation pattern (caused epigenetic alterations) at CpG sites within the region of DNA studied. In untreated cells, CpGs located at −189, −134, −29, −19, +16, and +19 of the BRCA1 gene had methylated cytosines (where numbering is from the messenger RNA transcription start site, and negative numbers indicate nucleotides in the upstream promoter region). Bromate treatment-induced oxidation resulted in the loss of cytosine methylation at −189, −134, +16 and +19 while also leading to the formation of new methylation at the CpGs located at −80, −55, −21 and +8 after DNA repair was allowed.\", 959)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Homologous recombinational repair alters epigenetic markers', 'At least four articles report the recruitment of DNA methyltransferase 1 (DNMT1) to sites of DNA double-strand breaks.  During homologous recombinational repair (HR) of the double-strand break, the involvement of DNMT1 causes the two repaired strands of DNA to have different levels of methylated cytosines.  One strand becomes frequently methylated at about 21 CpG sites downstream of the repaired double-strand break.  The other DNA strand loses methylation at about six CpG sites that were previously methylated downstream of the double-strand break, as well as losing methylation at about five CpG sites that were previously methylated upstream of the double-strand break.  When the chromosome is replicated, this gives rise to one daughter chromosome that is heavily methylated downstream of the previous break site and one that is unmethylated in the region both upstream and downstream of the previous break site.  With respect to the gene that was broken by the double-strand break, half of the progeny cells express that gene at a high level and in the other half of the progeny cells expression of that gene is repressed.  When clones of these cells were maintained for three years, the new methylation patterns were maintained over that time period.In mice with a CRISPR-mediated homology-directed recombination insertion in their genome there were a large number of increased methylations of CpG sites within the double-strand break-associated insertion.', 323)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Non-homologous end joining can cause some epigenetic marker alterations', 'Non-homologous end joining (NHEJ) repair of a double-strand break can cause a small number of demethylations of pre-existing cytosine DNA methylations downstream of the repaired double-strand break.  Further work by Allen et al. showed that NHEJ of a DNA double-strand break in a cell could give rise to some progeny cells having repressed expression of the gene harboring the initial double-strand break and some progeny having high expression of that gene due to epigenetic alterations associated with NHEJ repair.  The frequency of epigenetic alterations causing repression of a gene after an NHEJ repair of a DNA double-strand break in that gene may be about 0.9%.', 166)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Techniques used to study epigenetics', 'Epigenetic research uses a wide range of molecular biological techniques to further understanding of epigenetic phenomena. These techniques include chromatin immunoprecipitation (together with its large-scale variants ChIP-on-chip and ChIP-Seq), fluorescent in situ hybridization, methylation-sensitive restriction enzymes, DNA adenine methyltransferase identification (DamID) and bisulfite sequencing. Furthermore, the use of bioinformatics methods has a role in computational epigenetics.', 106)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Chromatin Immunoprecipitation', 'Chromatin Immunoprecipitation (ChIP) has helped bridge the gap between DNA and epigenetic interactions. With the use of ChIP, researchers are able to make findings in regards to gene regulation, transcription mechanisms, and chromatin structure.', 54)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Fluorescent in situ hybridization', 'Fluorescent in situ hybridization (FISH) is very important to understand epigenetic mechanisms. FISH can be used to find the location of genes on chromosomes, as well as finding noncoding RNAs. FISH is predominantly used for detecting chromosomal abnormalities in humans.', 62)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Methylation-sensitive restriction enzymes', 'Methylation sensitive restriction enzymes paired with PCR is a way to evaluate methylation in DNA - specifically the CpG sites. If DNA is methylated, the restriction enzymes will not cleave the strand. Contrarily, if the DNA is not methylated, the enzymes will cleave the strand and it will be amplified by PCR.', 73)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Bisulfite sequencing', 'Bisulfite sequencing is another way to evaluate DNA methylation. Cytosine will be changed to uracil from being treated with sodium bisulfite, whereas methylated cytosines will not be affected.', 48)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Mechanisms', 'Epigenetic mechanisms are sensitive to environmental effects, and key participants in shaping an adult phenotype. Several types of epigenetic inheritance systems may play a role in what has become known as cell memory, note however that not all of these are universally accepted to be examples of epigenetics.', 59)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Covalent modifications', 'Covalent modification of either DNA (e.g.. cytosine methylation and hydroxymethylation) or of histone proteins (e.g.. lysine acetylation, lysine and arginine methylation, serine and threonine phosphorylation, and lysine ubiquitination and sumoylation) play central roles in many types of epigenetic inheritance.. Therefore, the word \"epigenetics\" is sometimes used as a synonym for these processes.. However, this can be misleading.. Chromatin remodeling is not always inherited, and not all epigenetic inheritance involves chromatin remodeling.. In 2019, a further lysine modification appeared in the scientific literature linking epigenetics modification to cell metabolism, i.e.. Lactylation  Because the phenotype of a cell or individual is affected by which of its genes are transcribed, heritable transcription states can give rise to epigenetic effects.. There are several layers of regulation of gene expression.. One way that genes are regulated is through the remodeling of chromatin.. Chromatin is the complex of DNA and the histone proteins with which it associates.. If the way that DNA is wrapped around the histones changes, gene expression can change as well.. Chromatin remodeling is accomplished through two main mechanisms:  The first way is post translational modification of the amino acids that make up histone proteins.. Histone proteins are made up of long chains of amino acids.. If the amino acids that are in the chain are changed, the shape of the histone might be modified.. DNA is not completely unwound during replication.. It is possible, then, that the modified histones may be carried into each new copy of the DNA.. Once there, these histones may act as templates, initiating the surrounding new histones to be shaped in the new manner.. By altering the shape of the histones around them, these modified histones would ensure that a lineage-specific transcription program is maintained after cell division.. The second way is the addition of methyl groups to the DNA, mostly at CpG sites, to convert cytosine to 5-methylcytosine.. 5-Methylcytosine performs much like a regular cytosine, pairing with a guanine in double-stranded DNA.. However, when methylated cytosines are present in CpG sites in the promoter and enhancer regions of genes, the genes are often repressed..', 504)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'RNA transcripts', 'Sometimes a gene, after being turned on, transcribes a product that (directly or indirectly) maintains the activity of that gene. For example, Hnf4 and MyoD enhance the transcription of many liver-specific and muscle-specific genes, respectively, including their own, through the transcription factor activity of the proteins they encode. RNA signalling includes differential recruitment of a hierarchy of generic chromatin modifying complexes and DNA methyltransferases to specific loci by RNAs during differentiation and development. Other epigenetic changes are mediated by the production of different splice forms of RNA, or by formation of double-stranded RNA (RNAi). Descendants of the cell in which the gene was turned on will inherit this activity, even if the original stimulus for gene-activation is no longer present. These genes are often turned on or off by signal transduction, although in some systems where syncytia or gap junctions are important, RNA may spread directly to other cells or nuclei by diffusion. A large amount of RNA and protein is contributed to the zygote by the mother during oogenesis or via nurse cells, resulting in maternal effect phenotypes. A smaller quantity of sperm RNA is transmitted from the father, but there is recent evidence that this epigenetic information can lead to visible changes in several generations of offspring.', 272)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'MicroRNAs', 'MicroRNAs (miRNAs) are members of non-coding RNAs that range in size from 17 to 25 nucleotides. miRNAs regulate a large variety of biological functions in plants and animals. So far, in 2013, about 2000 miRNAs have been discovered in humans and these can be found online in a miRNA database. Each miRNA expressed in a cell may target about 100 to 200 messenger RNAs(mRNAs) that it downregulates. Most of the downregulation of mRNAs occurs by causing the decay of the targeted mRNA, while some downregulation occurs at the level of translation into protein.It appears that about 60% of human protein coding genes are regulated by miRNAs. Many miRNAs are epigenetically regulated. About 50% of miRNA genes are associated with CpG islands, that may be repressed by epigenetic methylation. Transcription from methylated CpG islands is strongly and heritably repressed. Other miRNAs are epigenetically regulated by either histone modifications or by combined DNA methylation and histone modification.', 228)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'mRNA', 'In 2011, it was demonstrated that the methylation of mRNA plays a critical role in human energy homeostasis. The obesity-associated FTO gene is shown to be able to demethylate N6-methyladenosine in RNA.', 52)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'sRNAs', 'sRNAs are small (50–250 nucleotides), highly structured, non-coding RNA fragments found in bacteria. They control gene expression including virulence genes in pathogens and are viewed as new targets in the fight against drug-resistant bacteria. They play an important role in many biological processes, binding to mRNA and protein targets in prokaryotes. Their phylogenetic analyses, for example through sRNA–mRNA target interactions or protein binding properties, are used to build comprehensive databases. sRNA-gene maps based on their targets in microbial genomes are also constructed.', 120)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Prions', \"Prions are infectious forms of proteins. In general, proteins fold into discrete units that perform distinct cellular functions, but some proteins are also capable of forming an infectious conformational state known as a prion. Although often viewed in the context of infectious disease, prions are more loosely defined by their ability to catalytically convert other native state versions of the same protein to an infectious conformational state. It is in this latter sense that they can be viewed as epigenetic agents capable of inducing a phenotypic change without a modification of the genome.Fungal prions are considered by some to be epigenetic because the infectious phenotype caused by the prion can be inherited without modification of the genome. PSI+ and URE3, discovered in yeast in 1965 and 1971, are the two best studied of this type of prion. Prions can have a phenotypic effect through the sequestration of protein in aggregates, thereby reducing that protein's activity. In PSI+ cells, the loss of the Sup35 protein (which is involved in termination of translation) causes ribosomes to have a higher rate of read-through of stop codons, an effect that results in suppression of nonsense mutations in other genes. The ability of Sup35 to form prions may be a conserved trait. It could confer an adaptive advantage by giving cells the ability to switch into a PSI+ state and express dormant genetic features normally terminated by stop codon mutations.Prion-based epigenetics has also been observed in Saccharomyces cerevisiae.\", 318)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Structural inheritance', 'In ciliates such as Tetrahymena and Paramecium, genetically identical cells show heritable differences in the patterns of ciliary rows on their cell surface. Experimentally altered patterns can be transmitted to daughter cells. It seems existing structures act as templates for new structures. The mechanisms of such inheritance are unclear, but reasons exist to assume that multicellular organisms also use existing cell structures to assemble new ones.', 88)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Nucleosome positioning', 'Eukaryotic genomes have numerous nucleosomes. Nucleosome position is not random, and determine the accessibility of DNA to regulatory proteins. Promoters active in different tissues have been shown to have different nucleosome positioning features. This determines differences in gene expression and cell differentiation. It has been shown that at least some nucleosomes are retained in sperm cells (where most but not all histones are replaced by protamines). Thus nucleosome positioning is to some degree inheritable. Recent studies have uncovered connections between nucleosome positioning and other epigenetic factors, such as DNA methylation and hydroxymethylation.', 131)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Histone variants', 'Different histone variants are incorporated into specific regions of the genome non-randomly. Their differential biochemical characteristics can affect genome functions via their roles in gene regulation, and maintenance of chromosome structures.', 41)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Genomic architecture', 'The three-dimensional configuration of the genome (the 3D genome) is complex, dynamic and crucial for regulating genomic function and nuclear processes such as DNA replication, transcription and DNA-damage repair.', 42)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Functions and consequences', '', 3)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'In the brain', '', 3)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Memory', 'Memory formation and maintenance are due to epigenetic alterations that cause the required dynamic changes in gene transcription that create and renew memory in neurons.An event can set off a chain of reactions that result in altered methylations of a large set of genes in neurons, which give a representation of the event, a memory.\\n\\nAreas of the brain important in the formation of memories include the hippocampus, medial prefrontal cortex (mPFC), anterior cingulate cortex and amygdala, as shown in the diagram of the human brain in this section.When a strong memory is created, as in a rat subjected to contextual fear conditioning (CFC), one of the earliest events to occur is that more than 100 DNA double-strand breaks are formed by topoisomerase IIB in neurons of the hippocampus and the medial prefrontal cortex (mPFC). These double-strand breaks are at specific locations that allow activation of transcription of immediate early genes (IEGs) that are important in memory formation, allowing their expression in mRNA, with peak mRNA transcription at seven to ten minutes after CFC.Two important IEGs in memory formation are EGR1 and the alternative promoter variant of DNMT3A, DNMT3A2.  EGR1 protein binds to DNA at its binding motifs, 5′-GCGTGGGCG-3′ or 5′-GCGGGGGCGG-3\\', and there are about 12,000 genome locations at which EGR1 protein can bind. EGR1 protein binds to DNA in gene promoter and enhancer regions.  EGR1 recruits the demethylating enzyme TET1 to an association, and brings TET1 to about 600 locations on the genome where TET1 can then demethylate and activate the associated genes.\\n\\nThe DNA methyltransferases DNMT3A1, DNMT3A2 and DNMT3B can all methylate cytosines (see image this section) at CpG sites in or near the promoters of genes. As shown by Manzo et al.,  these three DNA methyltransferases differ in their genomic binding locations and DNA methylation activity at different regulatory sites.  Manzo et al. located 3,970 genome regions exclusively enriched for DNMT3A1, 3,838 regions for DNMT3A2 and 3,432 regions for DNMT3B. When DNMT3A2 is newly induced as an IEG (when neurons are activated), many new cytosine methylations occur, presumably in the target regions of DNMT3A2. Oliviera et al. found that the neuronal activity-inducible IEG levels of Dnmt3a2 in the hippocampus determined the ability to form long-term memories.\\nRats form long-term associative memories after contextual fear conditioning (CFC). Duke et al. found that 24 hours after CFC in rats, in hippocampus neurons, 2,097 genes (9.17% of the genes in the rat genome) had altered methylation. When newly methylated cytosines are present in CpG sites in the promoter regions of genes, the genes are often repressed, and when newly demethylated cytosines are present the genes may be activated. After CFC, there were 1,048 genes with reduced mRNA expression and 564 genes with upregulated mRNA expression. Similarly, when mice undergo CFC, one hour later in the hippocampus region of the mouse brain there are 675 demethylated genes and 613 hypermethylated genes. However, memories do not remain in the hippocampus, but after four or five weeks the memories are stored in the anterior cingulate cortex. In the studies on mice after CFC, Halder et al. showed that four weeks after CFC there were at least 1,000 differentially methylated genes and more than 1,000 differentially expressed genes in the anterior cingulate cortex, while at the same time the altered methylations in the hippocampus were reversed.\\nThe epigenetic alteration of methylation after a new memory is established creates a different pool of nuclear mRNAs.  As reviewed by Bernstein, the epigenetically determined new mix of nuclear mRNAs are often packaged into neuronal granules, or messenger RNP, consisting of mRNA, small and large ribosomal subunits, translation initiation factors and RNA-binding proteins that regulate mRNA function. These neuronal granules are transported from the neuron nucleus and are directed, according to 3′ untranslated regions of the mRNA in the granules (their \"zip codes\"), to neuronal dendrites. Roughly 2,500 mRNAs may be localized to the dendrites of hippocampal pyramidal neurons and perhaps 450 transcripts are in excitatory presynaptic nerve terminals (dendritic spines). The altered assortments of transcripts (dependent on epigenetic alterations in the neuron nucleus) have different sensitivities in response to signals, which is the basis of altered synaptic plasticity. Altered synaptic plasticity is often considered the neurochemical foundation of learning and memory.', 1041)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Other and general', 'In adulthood, changes in the epigenome are important for various higher cognitive functions. Dysregulation of epigenetic mechanisms is implicated in neurodegenerative disorders and diseases. Epigenetic modifications in neurons are dynamic and reversible. Epigenetic regulation impacts neuronal action, affecting learning, memory, and other cognitive processes.Early events, including during embryonic development, can influence development, cognition, and health outcomes through epigenetic mechanisms.Epigenetic mechanisms have been proposed as \"a potential molecular mechanism for effects of endogenous hormones on the organization of developing brain circuits\".Nutrients could interact with the epigenome to \"protect or boost cognitive processes across the lifespan\".Epigenetics play a major role in brain aging and age-related cognitive decline, with relevance to life extension.A review suggests neurobiological effects of physical exercise via epigenetics seem \"central to building an \\'epigenetic memory\\' to influence long-term brain function and behavior\" and may even be heritable.With the axo-ciliary synapse, there is communication between serotonergic axons and antenna-like primary cilia of CA1 pyramidal neurons that alters the neuron\\'s epigenetic state in the  nucleus via the signalling distinct from that at the plasma membrane (and longer-term).Epigenetics also play a major role in the brain evolution in and to humans.', 276)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Development', 'Developmental epigenetics can be divided into predetermined and probabilistic epigenesis. Predetermined epigenesis is a unidirectional movement from structural development in DNA to the functional maturation of the protein. \"Predetermined\" here means that development is scripted and predictable. Probabilistic epigenesis on the other hand is a bidirectional structure-function development with experiences and external molding development.Somatic epigenetic inheritance, particularly through DNA and histone covalent modifications and nucleosome repositioning, is very important in the development of multicellular eukaryotic organisms. The genome sequence is static (with some notable exceptions), but cells differentiate into many different types, which perform different functions, and respond differently to the environment and intercellular signaling. Thus, as individuals develop, morphogens activate or silence genes in an epigenetically heritable fashion, giving cells a memory. In mammals, most cells terminally differentiate, with only stem cells retaining the ability to differentiate into several cell types (\"totipotency\" and \"multipotency\"). In mammals, some stem cells continue producing newly differentiated cells throughout life, such as in neurogenesis, but mammals are not able to respond to loss of some tissues, for example, the inability to regenerate limbs, which some other animals are capable of. Epigenetic modifications regulate the transition from neural stem cells to glial progenitor cells (for example, differentiation into oligodendrocytes is regulated by the deacetylation and methylation of histones. Unlike animals, plant cells do not terminally differentiate, remaining totipotent with the ability to give rise to a new individual plant. While plants do utilize many of the same epigenetic mechanisms as animals, such as chromatin remodeling, it has been hypothesized that some kinds of plant cells do not use or require \"cellular memories\", resetting their gene expression patterns using positional information from the environment and surrounding cells to determine their fate.Epigenetic changes can occur in response to environmental exposure – for example, maternal dietary supplementation with genistein (250 mg/kg) have epigenetic changes affecting expression of the agouti gene, which affects their fur color, weight, and propensity to develop cancer. Ongoing research is focused on exploring the impact of other known teratogens, such as diabetic embryopathy, on methylation signatures.Controversial results from one study suggested that traumatic experiences might produce an epigenetic signal that is capable of being passed to future generations. Mice were trained, using foot shocks, to fear a cherry blossom odor. The investigators reported that the mouse offspring had an increased aversion to this specific odor. They suggested epigenetic changes that increase gene expression, rather than in DNA itself, in a gene, M71, that governs the functioning of an odor receptor in the nose that responds specifically to this cherry blossom smell. There were physical changes that correlated with olfactory (smell) function in the brains of the trained mice and their descendants. Several criticisms were reported, including the study\\'s low statistical power as evidence of some irregularity such as bias in reporting results. Due to limits of sample size, there is a probability that an effect will not be demonstrated to within statistical significance even if it exists. The criticism suggested that the probability that all the experiments reported would show positive results if an identical protocol was followed, assuming the claimed effects exist, is merely 0.4%. The authors also did not indicate which mice were siblings, and treated all of the mice as statistically independent. The original researchers pointed out negative results in the paper\\'s appendix that the criticism omitted in its calculations, and undertook to track which mice were siblings in the future.', 754)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Transgenerational', 'Epigenetic mechanisms were a necessary part of the evolutionary origin of cell differentiation. Although epigenetics in multicellular organisms is generally thought to be a mechanism involved in differentiation, with epigenetic patterns \"reset\" when organisms reproduce, there have been some observations of transgenerational epigenetic inheritance (e.g., the phenomenon of paramutation observed in maize). Although most of these multigenerational epigenetic traits are gradually lost over several generations, the possibility remains that multigenerational epigenetics could be another aspect to evolution and adaptation.\\nAs mentioned above, some define epigenetics as heritable.\\nA sequestered germ line or Weismann barrier is specific to animals, and epigenetic inheritance is more common in plants and microbes. Eva Jablonka, Marion J. Lamb and Étienne Danchin have argued that these effects may require enhancements to the standard conceptual framework of the modern synthesis and have called for an extended evolutionary synthesis. Other evolutionary biologists, such as John Maynard Smith, have incorporated epigenetic inheritance into population-genetics models or are openly skeptical of the extended evolutionary synthesis (Michael Lynch). Thomas Dickins and Qazi Rahman state that epigenetic mechanisms such as DNA methylation and histone modification are genetically inherited under the control of natural selection and therefore fit under the earlier \"modern synthesis\".Two important ways in which epigenetic inheritance can differ from traditional genetic inheritance, with important consequences for evolution, are:\\n\\nrates of epimutation can be much faster than rates of mutation\\nthe epimutations are more easily reversibleIn plants, heritable DNA methylation mutations are 100,000 times more likely to occur compared to DNA mutations. An epigenetically inherited element such as the PSI+ system can act as a \"stop-gap\", good enough for short-term adaptation that allows the lineage to survive for long enough for mutation and/or recombination to genetically assimilate the adaptive phenotypic change. The existence of this possibility increases the evolvability of a species.\\nMore than 100 cases of transgenerational epigenetic inheritance phenomena have been reported in a wide range of organisms, including prokaryotes, plants, and animals. For instance, mourning-cloak butterflies will change color through hormone changes in response to experimentation of varying temperatures.The filamentous fungus Neurospora crassa is a prominent model system for understanding the control and function of cytosine methylation. In this organism, DNA methylation is associated with relics of a genome-defense system called RIP (repeat-induced point mutation) and silences gene expression by inhibiting transcription elongation.The yeast prion PSI is generated by a conformational change of a translation termination factor, which is then inherited by daughter cells. This can provide a survival advantage under adverse conditions, exemplifying epigenetic regulation which enables unicellular organisms to respond rapidly to environmental stress. Prions can be viewed as epigenetic agents capable of inducing a phenotypic change without modification of the genome.Direct detection of epigenetic marks in microorganisms is possible with single molecule real time sequencing, in which polymerase sensitivity allows for measuring methylation and other modifications as a DNA molecule is being sequenced. Several projects have demonstrated the ability to collect genome-wide epigenetic data in bacteria.', 667)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Epigenetics in bacteria', 'While epigenetics is of fundamental importance in eukaryotes, especially metazoans, it plays a different role in bacteria. Most importantly, eukaryotes use epigenetic mechanisms primarily to regulate gene expression which bacteria rarely do. However, bacteria make widespread use of postreplicative DNA methylation for the epigenetic control of DNA-protein interactions. Bacteria also use DNA adenine methylation (rather than DNA cytosine methylation) as an epigenetic signal. DNA adenine methylation is important in bacteria virulence in organisms such as Escherichia coli, Salmonella, Vibrio, Yersinia, Haemophilus, and Brucella. In Alphaproteobacteria, methylation of adenine regulates the cell cycle and couples gene transcription to DNA replication. In Gammaproteobacteria, adenine methylation provides signals for DNA replication, chromosome segregation, mismatch repair, packaging of bacteriophage, transposase activity and regulation of gene expression. There exists a genetic switch controlling Streptococcus pneumoniae (the pneumococcus) that allows the bacterium to randomly change its characteristics into six alternative states that could pave the way to improved vaccines. Each form is randomly generated by a phase variable methylation system. The ability of the pneumococcus to cause deadly infections is different in each of these six states. Similar systems exist in other bacterial genera. In Bacillota such as Clostridioides difficile, adenine methylation regulates sporulation, biofilm formation and host-adaptation.', 328)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Medicine', 'Epigenetics has many and varied potential medical applications. In 2008, the National Institutes of Health announced that $190 million had been earmarked for epigenetics research over the next five years. In announcing the funding, government officials noted that epigenetics has the potential to explain mechanisms of aging, human development, and the origins of cancer, heart disease, mental illness, as well as several other conditions. Some investigators, like Randy Jirtle, Ph.D., of Duke University Medical Center, think epigenetics may ultimately turn out to have a greater role in disease than genetics.', 119)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Twins', 'Direct comparisons of identical twins constitute an optimal model for interrogating environmental epigenetics. In the case of humans with different environmental exposures, monozygotic (identical) twins were epigenetically indistinguishable during their early years, while older twins had remarkable differences in the overall content and genomic distribution of 5-methylcytosine DNA and histone acetylation. The twin pairs who had spent less of their lifetime together and/or had greater differences in their medical histories were those who showed the largest differences in their levels of 5-methylcytosine DNA and acetylation of histones H3 and H4.Dizygotic (fraternal) and monozygotic (identical) twins show evidence of epigenetic influence in humans. DNA sequence differences that would be abundant in a singleton-based study do not interfere with the analysis. Environmental differences can produce long-term epigenetic effects, and different developmental monozygotic twin subtypes may be different with respect to their susceptibility to be discordant from an epigenetic point of view.A high-throughput study, which denotes technology that looks at extensive genetic markers, focused on epigenetic differences between monozygotic twins to compare global and locus-specific changes in DNA methylation and histone modifications in a sample of 40 monozygotic twin pairs. In this case, only healthy twin pairs were studied, but a wide range of ages was represented, between 3 and 74 years. One of the major conclusions from this study was that there is an age-dependent accumulation of epigenetic differences between the two siblings of twin pairs. This accumulation suggests the existence of epigenetic \"drift\". Epigenetic drift is the term given to epigenetic modifications as they occur as a direct function with age. While age is a known risk factor for many diseases, age-related methylation\\nhas been found to occur differentially at specific sites along the genome. Over time, this can result in measurable differences between biological and chronological age. Epigenetic changes have been found to be reflective of lifestyle and may act as functional biomarkers of disease before clinical threshold is reached.A more recent study, where 114 monozygotic twins and 80 dizygotic twins were analyzed for the DNA methylation status of around 6000 unique genomic regions, concluded that epigenetic similarity at the time of blastocyst splitting may also contribute to phenotypic similarities in monozygotic co-twins. This supports the notion that microenvironment at early stages of embryonic development can be quite important for the establishment of epigenetic marks.\\nCongenital genetic disease is well understood and it is clear that epigenetics can play a role, for example, in the case of Angelman syndrome and Prader-Willi syndrome. These are normal genetic diseases caused by gene deletions or inactivation of the genes but are unusually common because individuals are essentially hemizygous because of genomic imprinting, and therefore a single gene knock out is sufficient to cause the disease, where most cases would require both copies to be knocked out.', 619)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Genomic imprinting', \"Some human disorders are associated with genomic imprinting, a phenomenon in mammals where the father and mother contribute different epigenetic patterns for specific genomic loci in their germ cells. The best-known case of imprinting in human disorders is that of Angelman syndrome and Prader-Willi syndrome – both can be produced by the same genetic mutation, chromosome 15q partial deletion, and the particular syndrome that will develop depends on whether the mutation is inherited from the child's mother or from their father. This is due to the presence of genomic imprinting in the region. Beckwith-Wiedemann syndrome is also associated with genomic imprinting, often caused by abnormalities in maternal genomic imprinting of a region on chromosome 11.\\nMethyl CpG-binding protein 2 (MeCP2) is a transcriptional regulator that must be phosphorylated before releasing from the BDNF promoter, allowing transcription. Rett syndrome is underlain by mutations in the MeCP2 gene despite no large-scale changes in expression of MeCP2 being found in microarray analyses. BDNF is downregulated in the MECP2 mutant resulting in Rett syndrome, as well as the increase of early neural senescence and accumulation of damaged DNA.In the Överkalix study, paternal (but not maternal) grandsons of Swedish men who were exposed during preadolescence to famine in the 19th century were less likely to die of cardiovascular disease. If food was plentiful, then diabetes mortality in the grandchildren increased, suggesting that this was a transgenerational epigenetic inheritance. The opposite effect was observed for females – the paternal (but not maternal) granddaughters of women who experienced famine while in the womb (and therefore while their eggs were being formed) lived shorter lives on average.\", 366)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Diabetic wound healing', 'Epigenetic modifications have given insight into the understanding of the pathophysiology of different disease conditions. Though, they are strongly associated with cancer, their role in other pathological conditions are of equal importance. It appears that the hyperglycaemic environment could imprint such changes at the genomic level, that macrophages are primed towards a pro-inflammatory state and could fail to exhibit any phenotypic alteration towards the pro-healing type. This phenomenon of altered Macrophage Polarization is mostly associated with all the diabetic complications in a clinical set-up. As of 2018, several reports reveal the relevance of different epigenetic modifications with respect to diabetic complications. Sooner or later, with the advancements in biomedical tools, the detection of such biomarkers as prognostic and diagnostic tools in patients could possibly emerge out as alternative approaches. It is noteworthy to mention here that the use of epigenetic modifications as therapeutic targets warrant extensive preclinical as well as clinical evaluation prior to use.', 200)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Examples of drugs altering gene expression from epigenetic events', 'The use of beta-lactam antibiotics can alter glutamate receptor activity and the action of cyclosporine on multiple transcription factors. Additionally, lithium can impact autophagy of aberrant proteins, and opioid drugs via chronic use can increase the expression of genes associated with addictive phenotypes.In a groundbreaking 2003 report, Caspi and colleagues demonstrated that in a robust cohort of over one-thousand subjects assessed multiple times from preschool to adulthood, subjects who carried one or two copies of the short allele of the serotonin transporter promoter polymorphism exhibited higher rates of adult depression and suicidality when exposed to childhood maltreatment when compared to long allele homozygotes with equal ELS exposure.Parental nutrition, in utero exposure to stress or endocrine disrupting chemicals, male-induced maternal effects such as the attraction of differential mate quality, and maternal as well as paternal age, and offspring gender could all possibly influence whether a germline epimutation is ultimately expressed in offspring and the degree to which intergenerational inheritance remains stable throughout posterity. However, whether and to what extent epigenetic effects can be transmitted across generations remains unclear, particularly in humans.', 244)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Addiction', \"Addiction is a disorder of the brain's reward system which arises through transcriptional and neuroepigenetic mechanisms and occurs over time from chronically high levels of exposure to an addictive stimulus (e.g., morphine, cocaine, sexual intercourse, gambling, etc.). Transgenerational epigenetic inheritance of addictive phenotypes has been noted to occur in preclinical studies. However, robust evidence in support of the persistence of epigenetic effects across multiple generations has yet to be established in humans; for example, an epigenetic effect of prenatal exposure to smoking that is observed in great-grandchildren who had not been exposed.\", 125)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Depression', 'Epigenetic inheritance of depression-related phenotypes has also been reported in a preclinical study. Inheritance of paternal stress-induced traits across generations involved small non-coding RNA signals transmitted via the paternal germline.', 48)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Research', 'The two forms of heritable information, namely genetic and epigenetic, are collectively called dual inheritance. Members of the APOBEC/AID family of cytosine deaminases may concurrently influence genetic and epigenetic inheritance using similar molecular mechanisms, and may be a point of crosstalk between these conceptually compartmentalized processes.Fluoroquinolone antibiotics induce epigenetic changes in mammalian cells through iron chelation. This leads to epigenetic effects through inhibition of α-ketoglutarate-dependent dioxygenases that require iron as a co-factor.Various pharmacological agents are applied for the production of induced pluripotent stem cells (iPSC) or maintain the embryonic stem cell (ESC) phenotypic via epigenetic approach. Adult stem cells like bone marrow stem cells have also shown a potential to differentiate into cardiac competent cells when treated with G9a histone methyltransferase inhibitor BIX01294.', 197)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Epigenome editing', 'Epigenetic regulation of gene expression that could be altered or used in epigenome editing are or include mRNA/lncRNA modification, DNA methylation modification and histone modification.', 39)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'CpG sites, SNPs and biological traits', 'Methylation is a widely characterized mechanism of genetic regulation that can determine biological traits. However, strong experimental evidences correlate methylation patterns in SNPs as an important additional feature for the classical activation/inhibition epigenetic dogma. Molecular interaction data, supported by colocalization analyses, identify multiple nuclear regulatory pathways, linking sequence variation to disturbances in DNA methylation and molecular and phenotypic variation.', 89)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'UBASH3B locus', 'UBASH3B encodes a protein with tyrosine phosphatase activity, which has been previously linked to advanced neoplasia. SNP rs7115089 was identified as influencing DNA methylation and expression of this locus, as well as and Body Mass Index (BMI). In fact, SNP rs7115089 is strongly associated with BMI and with genetic variants linked to other cardiovascular and metabolic traits in GWASs. New studies suggesting UBASH3B as a potential mediator of adiposity and cardiometabolic disease. In addition, animal models demonstrated that UBASH3B expression is an indicator of caloric restriction that may drive programmed susceptibility to obesity and it is associated with other measures of adiposity in human peripherical blood.', 157)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'NFKBIE locus', 'SNP rs730775 is located in the first intron of NFKBIE and is a cis eQTL for NFKBIE in whole blood. Nuclear factor (NF)-κB inhibitor ε (NFKBIE) directly inhibits NF-κB1 activity and is significantly co-expressed with NF-κB1, also, it is associated with rheumatoid arthritis. Colocalization analysis supports that variants for the majority of the CpG sites in SNP rs730775 cause genetic variation at the NFKBIE locus which is suggestable linked to rheumatoid arthritis through trans acting regulation of DNA methylation by NF-κB.', 145)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'FADS1 locus', 'Fatty acid desaturase 1 (FADS1) is a key enzyme in the metabolism of fatty acids. Moreover, rs174548 in the FADS1 gene shows increased correlation with DNA methylation in people with high abundance of CD8+ T cells. SNP rs174548 is strongly associated with concentrations of arachidonic acid and other metabolites in fatty acid metabolism, blood eosinophil counts. and inflammatory diseases such as asthma. Interaction results indicated a correlation between rs174548 and asthma, providing new insights about fatty acid metabolism in CD8+ T cells with immune phenotypes.', 128)\n",
      "\n",
      "Section:\n",
      "\n",
      "(1136190969, 'Epigenetics', 'Pseudoscience', \"As epigenetics is in the early stages of development as a science and is surrounded by sensationalism in the public media, David Gorski and geneticist Adam Rutherford have advised caution against the proliferation of false and pseudoscientific conclusions by new age authors making unfounded suggestions that a person's genes and health can be manipulated by mind control. Misuse of the scientific term by quack authors has produced misinformation among the general public.\", 88)\n"
     ]
    }
   ],
   "source": [
    "# Example page being processed into sections\n",
    "\n",
    "discard_categories = ['See also', 'References', 'External links', 'Further reading', \"Footnotes\",\n",
    "    \"Bibliography\", \"Sources\", \"Citations\", \"Literature\", \"Footnotes\", \"Notes and references\",\n",
    "    \"Photo gallery\", \"Works cited\", \"Photos\", \"Gallery\", \"Notes\", \"References and sources\",\n",
    "    \"References and notes\"]\n",
    "\n",
    "example_title = 'Epigenomics'\n",
    "example_page = get_wiki_page(example_title)\n",
    "sections = extract_sections(example_page.content, example_page.revision_id, example_page.title, discard_categories=discard_categories)\n",
    "\n",
    "print(f\"There are {len(sections)} sections in the page {example_title}\")\n",
    "for section in sections:\n",
    "    print('\\nSection:\\n')\n",
    "    print(section)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "c7f77deb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>page_id</th>\n",
       "      <th>title</th>\n",
       "      <th>heading</th>\n",
       "      <th>content</th>\n",
       "      <th>tokens</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>Etymology</td>\n",
       "      <td>From the Greek ΓΕΝ gen, \"gene\" (gamma, epsilon...</td>\n",
       "      <td>166.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1137582695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>Early sequencing efforts</td>\n",
       "      <td>Following Rosalind Franklin's confirmation of ...</td>\n",
       "      <td>248.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1137582695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>DNA-sequencing technology developed</td>\n",
       "      <td>In addition to his seminal work on the amino a...</td>\n",
       "      <td>355.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1137582695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>Complete genomes</td>\n",
       "      <td>The advent of these technologies resulted in a...</td>\n",
       "      <td>691.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1137582695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>The \"omics\" revolution</td>\n",
       "      <td>The English-language neologism omics informall...</td>\n",
       "      <td>198.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1137582695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308</th>\n",
       "      <td>1060657820</td>\n",
       "      <td>Toxicogenomics</td>\n",
       "      <td>Public projects</td>\n",
       "      <td>Chemical Effects in Biological Systems is a pr...</td>\n",
       "      <td>307.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1060657820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1309</th>\n",
       "      <td>1061709342</td>\n",
       "      <td>Comparative Toxicogenomics Database</td>\n",
       "      <td>Background</td>\n",
       "      <td>The Comparative Toxicogenomics Database (CTD) ...</td>\n",
       "      <td>79.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1061709342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1310</th>\n",
       "      <td>1061709342</td>\n",
       "      <td>Comparative Toxicogenomics Database</td>\n",
       "      <td>Goals and objectives</td>\n",
       "      <td>One of the primary goals of CTD is to advance ...</td>\n",
       "      <td>144.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1061709342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1311</th>\n",
       "      <td>1061709342</td>\n",
       "      <td>Comparative Toxicogenomics Database</td>\n",
       "      <td>Core data</td>\n",
       "      <td>CTD is a unique resource where biocurators rea...</td>\n",
       "      <td>58.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1061709342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312</th>\n",
       "      <td>1061709342</td>\n",
       "      <td>Comparative Toxicogenomics Database</td>\n",
       "      <td>Data integration</td>\n",
       "      <td>By integrating the above four data sets, CTD a...</td>\n",
       "      <td>250.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1061709342</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1313 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         page_id                                title  \\\n",
       "0     1137582695                             Genomics   \n",
       "1     1137582695                             Genomics   \n",
       "2     1137582695                             Genomics   \n",
       "3     1137582695                             Genomics   \n",
       "4     1137582695                             Genomics   \n",
       "...          ...                                  ...   \n",
       "1308  1060657820                       Toxicogenomics   \n",
       "1309  1061709342  Comparative Toxicogenomics Database   \n",
       "1310  1061709342  Comparative Toxicogenomics Database   \n",
       "1311  1061709342  Comparative Toxicogenomics Database   \n",
       "1312  1061709342  Comparative Toxicogenomics Database   \n",
       "\n",
       "                                  heading  \\\n",
       "0                               Etymology   \n",
       "1                Early sequencing efforts   \n",
       "2     DNA-sequencing technology developed   \n",
       "3                        Complete genomes   \n",
       "4                  The \"omics\" revolution   \n",
       "...                                   ...   \n",
       "1308                      Public projects   \n",
       "1309                           Background   \n",
       "1310                 Goals and objectives   \n",
       "1311                            Core data   \n",
       "1312                     Data integration   \n",
       "\n",
       "                                                content  tokens  \\\n",
       "0     From the Greek ΓΕΝ gen, \"gene\" (gamma, epsilon...   166.0   \n",
       "1     Following Rosalind Franklin's confirmation of ...   248.0   \n",
       "2     In addition to his seminal work on the amino a...   355.0   \n",
       "3     The advent of these technologies resulted in a...   691.0   \n",
       "4     The English-language neologism omics informall...   198.0   \n",
       "...                                                 ...     ...   \n",
       "1308  Chemical Effects in Biological Systems is a pr...   307.0   \n",
       "1309  The Comparative Toxicogenomics Database (CTD) ...    79.0   \n",
       "1310  One of the primary goals of CTD is to advance ...   144.0   \n",
       "1311  CTD is a unique resource where biocurators rea...    58.0   \n",
       "1312  By integrating the above four data sets, CTD a...   250.0   \n",
       "\n",
       "                                             url  \n",
       "0     https://en.wikipedia.org/?oldid=1137582695  \n",
       "1     https://en.wikipedia.org/?oldid=1137582695  \n",
       "2     https://en.wikipedia.org/?oldid=1137582695  \n",
       "3     https://en.wikipedia.org/?oldid=1137582695  \n",
       "4     https://en.wikipedia.org/?oldid=1137582695  \n",
       "...                                          ...  \n",
       "1308  https://en.wikipedia.org/?oldid=1060657820  \n",
       "1309  https://en.wikipedia.org/?oldid=1061709342  \n",
       "1310  https://en.wikipedia.org/?oldid=1061709342  \n",
       "1311  https://en.wikipedia.org/?oldid=1061709342  \n",
       "1312  https://en.wikipedia.org/?oldid=1061709342  \n",
       "\n",
       "[1313 rows x 6 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Build a data set from the wikipedia page sections\n",
    "\n",
    "res = []\n",
    "page_ids = []\n",
    "for page in pages:\n",
    "    res += extract_sections(page.content, page.revision_id, page.title, discard_categories=discard_categories)\n",
    "    page_ids.append(page.revision_id)\n",
    "df = pd.DataFrame(res, columns=[\"page_id\", \"title\", \"heading\", \"content\", \"tokens\"])\n",
    "df['url'] = df['page_id'].apply(lambda v: f\"https://en.wikipedia.org/?oldid={v}\")\n",
    "df = df[df.tokens>40]\n",
    "df = df.drop_duplicates(['title','heading'])\n",
    "df = df.reset_index().drop('index',axis=1) # reset index\n",
    "df.head()\n",
    "\n",
    "df.to_csv('genomics-data/genomics_sections.csv', index=False)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "daeb947d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://en.wikipedia.org/?oldid=0       1137582695\\n1       1137582695\\n2       1137582695\\n3       1137582695\\n4       1137582695\\n           ...    \\n1743    1061709342\\n1744    1061709342\\n1745    1061709342\\n1746    1061709342\\n1747    1061709342\\nName: page_id, Length: 1748, dtype: int64'"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = df['url'].unique()\n",
    "a[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "bee0a8c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>page_id</th>\n",
       "      <th>title</th>\n",
       "      <th>heading</th>\n",
       "      <th>content</th>\n",
       "      <th>tokens</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>Etymology</td>\n",
       "      <td>From the Greek ΓΕΝ gen, \"gene\" (gamma, epsilon...</td>\n",
       "      <td>166.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1137582695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>Early sequencing efforts</td>\n",
       "      <td>Following Rosalind Franklin's confirmation of ...</td>\n",
       "      <td>248.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1137582695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>DNA-sequencing technology developed</td>\n",
       "      <td>In addition to his seminal work on the amino a...</td>\n",
       "      <td>355.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1137582695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>Complete genomes</td>\n",
       "      <td>The advent of these technologies resulted in a...</td>\n",
       "      <td>691.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1137582695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>The \"omics\" revolution</td>\n",
       "      <td>The English-language neologism omics informall...</td>\n",
       "      <td>198.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1137582695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308</th>\n",
       "      <td>1060657820</td>\n",
       "      <td>Toxicogenomics</td>\n",
       "      <td>Public projects</td>\n",
       "      <td>Chemical Effects in Biological Systems is a pr...</td>\n",
       "      <td>307.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1060657820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1309</th>\n",
       "      <td>1061709342</td>\n",
       "      <td>Comparative Toxicogenomics Database</td>\n",
       "      <td>Background</td>\n",
       "      <td>The Comparative Toxicogenomics Database (CTD) ...</td>\n",
       "      <td>79.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1061709342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1310</th>\n",
       "      <td>1061709342</td>\n",
       "      <td>Comparative Toxicogenomics Database</td>\n",
       "      <td>Goals and objectives</td>\n",
       "      <td>One of the primary goals of CTD is to advance ...</td>\n",
       "      <td>144.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1061709342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1311</th>\n",
       "      <td>1061709342</td>\n",
       "      <td>Comparative Toxicogenomics Database</td>\n",
       "      <td>Core data</td>\n",
       "      <td>CTD is a unique resource where biocurators rea...</td>\n",
       "      <td>58.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1061709342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312</th>\n",
       "      <td>1061709342</td>\n",
       "      <td>Comparative Toxicogenomics Database</td>\n",
       "      <td>Data integration</td>\n",
       "      <td>By integrating the above four data sets, CTD a...</td>\n",
       "      <td>250.0</td>\n",
       "      <td>https://en.wikipedia.org/?oldid=1061709342</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1313 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         page_id                                title  \\\n",
       "0     1137582695                             Genomics   \n",
       "1     1137582695                             Genomics   \n",
       "2     1137582695                             Genomics   \n",
       "3     1137582695                             Genomics   \n",
       "4     1137582695                             Genomics   \n",
       "...          ...                                  ...   \n",
       "1308  1060657820                       Toxicogenomics   \n",
       "1309  1061709342  Comparative Toxicogenomics Database   \n",
       "1310  1061709342  Comparative Toxicogenomics Database   \n",
       "1311  1061709342  Comparative Toxicogenomics Database   \n",
       "1312  1061709342  Comparative Toxicogenomics Database   \n",
       "\n",
       "                                  heading  \\\n",
       "0                               Etymology   \n",
       "1                Early sequencing efforts   \n",
       "2     DNA-sequencing technology developed   \n",
       "3                        Complete genomes   \n",
       "4                  The \"omics\" revolution   \n",
       "...                                   ...   \n",
       "1308                      Public projects   \n",
       "1309                           Background   \n",
       "1310                 Goals and objectives   \n",
       "1311                            Core data   \n",
       "1312                     Data integration   \n",
       "\n",
       "                                                content  tokens  \\\n",
       "0     From the Greek ΓΕΝ gen, \"gene\" (gamma, epsilon...   166.0   \n",
       "1     Following Rosalind Franklin's confirmation of ...   248.0   \n",
       "2     In addition to his seminal work on the amino a...   355.0   \n",
       "3     The advent of these technologies resulted in a...   691.0   \n",
       "4     The English-language neologism omics informall...   198.0   \n",
       "...                                                 ...     ...   \n",
       "1308  Chemical Effects in Biological Systems is a pr...   307.0   \n",
       "1309  The Comparative Toxicogenomics Database (CTD) ...    79.0   \n",
       "1310  One of the primary goals of CTD is to advance ...   144.0   \n",
       "1311  CTD is a unique resource where biocurators rea...    58.0   \n",
       "1312  By integrating the above four data sets, CTD a...   250.0   \n",
       "\n",
       "                                             url  \n",
       "0     https://en.wikipedia.org/?oldid=1137582695  \n",
       "1     https://en.wikipedia.org/?oldid=1137582695  \n",
       "2     https://en.wikipedia.org/?oldid=1137582695  \n",
       "3     https://en.wikipedia.org/?oldid=1137582695  \n",
       "4     https://en.wikipedia.org/?oldid=1137582695  \n",
       "...                                          ...  \n",
       "1308  https://en.wikipedia.org/?oldid=1060657820  \n",
       "1309  https://en.wikipedia.org/?oldid=1061709342  \n",
       "1310  https://en.wikipedia.org/?oldid=1061709342  \n",
       "1311  https://en.wikipedia.org/?oldid=1061709342  \n",
       "1312  https://en.wikipedia.org/?oldid=1061709342  \n",
       "\n",
       "[1313 rows x 6 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['url'] = df['page_id'].apply(lambda v: f\"https://en.wikipedia.org/?oldid={v}\")\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a194b772",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>url</th>\n",
       "      <th>title</th>\n",
       "      <th>heading</th>\n",
       "      <th>content</th>\n",
       "      <th>tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>Etymology</td>\n",
       "      <td>From the Greek ΓΕΝ gen, \"gene\" (gamma, epsilon...</td>\n",
       "      <td>166.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>Early sequencing efforts</td>\n",
       "      <td>Following Rosalind Franklin's confirmation of ...</td>\n",
       "      <td>248.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>DNA-sequencing technology developed</td>\n",
       "      <td>In addition to his seminal work on the amino a...</td>\n",
       "      <td>355.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>Complete genomes</td>\n",
       "      <td>The advent of these technologies resulted in a...</td>\n",
       "      <td>691.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1137582695</td>\n",
       "      <td>Genomics</td>\n",
       "      <td>The \"omics\" revolution</td>\n",
       "      <td>The English-language neologism omics informall...</td>\n",
       "      <td>198.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1308</th>\n",
       "      <td>1060657820</td>\n",
       "      <td>Toxicogenomics</td>\n",
       "      <td>Public projects</td>\n",
       "      <td>Chemical Effects in Biological Systems is a pr...</td>\n",
       "      <td>307.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1309</th>\n",
       "      <td>1061709342</td>\n",
       "      <td>Comparative Toxicogenomics Database</td>\n",
       "      <td>Background</td>\n",
       "      <td>The Comparative Toxicogenomics Database (CTD) ...</td>\n",
       "      <td>79.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1310</th>\n",
       "      <td>1061709342</td>\n",
       "      <td>Comparative Toxicogenomics Database</td>\n",
       "      <td>Goals and objectives</td>\n",
       "      <td>One of the primary goals of CTD is to advance ...</td>\n",
       "      <td>144.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1311</th>\n",
       "      <td>1061709342</td>\n",
       "      <td>Comparative Toxicogenomics Database</td>\n",
       "      <td>Core data</td>\n",
       "      <td>CTD is a unique resource where biocurators rea...</td>\n",
       "      <td>58.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1312</th>\n",
       "      <td>1061709342</td>\n",
       "      <td>Comparative Toxicogenomics Database</td>\n",
       "      <td>Data integration</td>\n",
       "      <td>By integrating the above four data sets, CTD a...</td>\n",
       "      <td>250.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1313 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             url                                title  \\\n",
       "0     1137582695                             Genomics   \n",
       "1     1137582695                             Genomics   \n",
       "2     1137582695                             Genomics   \n",
       "3     1137582695                             Genomics   \n",
       "4     1137582695                             Genomics   \n",
       "...          ...                                  ...   \n",
       "1308  1060657820                       Toxicogenomics   \n",
       "1309  1061709342  Comparative Toxicogenomics Database   \n",
       "1310  1061709342  Comparative Toxicogenomics Database   \n",
       "1311  1061709342  Comparative Toxicogenomics Database   \n",
       "1312  1061709342  Comparative Toxicogenomics Database   \n",
       "\n",
       "                                  heading  \\\n",
       "0                               Etymology   \n",
       "1                Early sequencing efforts   \n",
       "2     DNA-sequencing technology developed   \n",
       "3                        Complete genomes   \n",
       "4                  The \"omics\" revolution   \n",
       "...                                   ...   \n",
       "1308                      Public projects   \n",
       "1309                           Background   \n",
       "1310                 Goals and objectives   \n",
       "1311                            Core data   \n",
       "1312                     Data integration   \n",
       "\n",
       "                                                content  tokens  \n",
       "0     From the Greek ΓΕΝ gen, \"gene\" (gamma, epsilon...   166.0  \n",
       "1     Following Rosalind Franklin's confirmation of ...   248.0  \n",
       "2     In addition to his seminal work on the amino a...   355.0  \n",
       "3     The advent of these technologies resulted in a...   691.0  \n",
       "4     The English-language neologism omics informall...   198.0  \n",
       "...                                                 ...     ...  \n",
       "1308  Chemical Effects in Biological Systems is a pr...   307.0  \n",
       "1309  The Comparative Toxicogenomics Database (CTD) ...    79.0  \n",
       "1310  One of the primary goals of CTD is to advance ...   144.0  \n",
       "1311  CTD is a unique resource where biocurators rea...    58.0  \n",
       "1312  By integrating the above four data sets, CTD a...   250.0  \n",
       "\n",
       "[1313 rows x 5 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.apply(lambda s: f\"https://en.wikipedia.org/?oldid={s['url']}\" if s.name == 'url' else s, axis=1)\n",
    "\n",
    "#df.apply(lambda x: np.square(x) if x.name == 'd' else x, axis=1)\n",
    "#df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4b0d28bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>url</th>\n",
       "      <th>title</th>\n",
       "      <th>heading</th>\n",
       "      <th>content</th>\n",
       "      <th>tokens</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>695</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>List of sequenced animal genomes</td>\n",
       "      <td>Bony fish</td>\n",
       "      <td>Order Anabantiformes\\nBetta splendens, Siamese...</td>\n",
       "      <td>1496.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>Neanderthal genome project</td>\n",
       "      <td>Findings</td>\n",
       "      <td>The researchers recovered ancient DNA of Neand...</td>\n",
       "      <td>1481.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1210</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>Biological effects of radiation on the epigenome</td>\n",
       "      <td>The lethal-potentially lethal model</td>\n",
       "      <td>This equation explores the hypothesis of a les...</td>\n",
       "      <td>1464.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1227</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>Gene</td>\n",
       "      <td>Conflicting definitions of 'gene'</td>\n",
       "      <td>There are lots of different ways to use the te...</td>\n",
       "      <td>1294.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>DeCODE genetics</td>\n",
       "      <td>Discoveries and scientific contributions</td>\n",
       "      <td>Genome research in general, and deCODE's globa...</td>\n",
       "      <td>1277.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>Clinical metagenomic sequencing</td>\n",
       "      <td>Laboratory workflow</td>\n",
       "      <td>A typical mNGS workflow consists of the follow...</td>\n",
       "      <td>1275.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>DeCODE genetics</td>\n",
       "      <td>Genome of a Nation</td>\n",
       "      <td>By the time Bill Clinton and Tony Blair announ...</td>\n",
       "      <td>1268.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>Rat Genome Database</td>\n",
       "      <td>Data</td>\n",
       "      <td>RGD’s data consists of manual annotations from...</td>\n",
       "      <td>1239.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>Neanderthal genetics</td>\n",
       "      <td>Interbreeding with anatomically modern humans</td>\n",
       "      <td>Researchers addressed the question of possible...</td>\n",
       "      <td>1237.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>Economics</td>\n",
       "      <td>Definitions of economics over time</td>\n",
       "      <td>The earlier term for the discipline was 'polit...</td>\n",
       "      <td>1231.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>DeCODE genetics</td>\n",
       "      <td>Product innovation</td>\n",
       "      <td>deCODE's scientific leadership over more than ...</td>\n",
       "      <td>1209.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>Economics</td>\n",
       "      <td>Neoclassical economics</td>\n",
       "      <td>At the dawn as a social science, economics was...</td>\n",
       "      <td>1188.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1156</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>Vertebrate Genome Annotation Project</td>\n",
       "      <td>Comparative analysis</td>\n",
       "      <td>In addition to full genomes, and unlike other ...</td>\n",
       "      <td>1182.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>Biological effects of radiation on the epigenome</td>\n",
       "      <td>The three lambda model</td>\n",
       "      <td>This model showed the accuracy of survival des...</td>\n",
       "      <td>1129.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>Rat Genome Database</td>\n",
       "      <td>Genome tools developed at RGD</td>\n",
       "      <td>RGD develops web-based tools designed to use t...</td>\n",
       "      <td>1106.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1082</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>Population genetics</td>\n",
       "      <td>Genetic drift</td>\n",
       "      <td>Genetic drift is a change in allele frequencie...</td>\n",
       "      <td>1103.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1027</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>Genome editing</td>\n",
       "      <td>Zinc finger nucleases</td>\n",
       "      <td>As opposed to meganucleases, the concept behin...</td>\n",
       "      <td>1072.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>709</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>BGI Group</td>\n",
       "      <td>History</td>\n",
       "      <td>Wang Jian, Yu Jun, Yang Huanming and Liu Siqi ...</td>\n",
       "      <td>1071.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1183</th>\n",
       "      <td>https://en.wikipedia.org/?oldid=0       113758...</td>\n",
       "      <td>Epigenetics</td>\n",
       "      <td>Memory</td>\n",
       "      <td>Memory formation and maintenance are due to ep...</td>\n",
       "      <td>1041.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    url  \\\n",
       "695   https://en.wikipedia.org/?oldid=0       113758...   \n",
       "226   https://en.wikipedia.org/?oldid=0       113758...   \n",
       "1210  https://en.wikipedia.org/?oldid=0       113758...   \n",
       "1227  https://en.wikipedia.org/?oldid=0       113758...   \n",
       "96    https://en.wikipedia.org/?oldid=0       113758...   \n",
       "161   https://en.wikipedia.org/?oldid=0       113758...   \n",
       "95    https://en.wikipedia.org/?oldid=0       113758...   \n",
       "449   https://en.wikipedia.org/?oldid=0       113758...   \n",
       "230   https://en.wikipedia.org/?oldid=0       113758...   \n",
       "173   https://en.wikipedia.org/?oldid=0       113758...   \n",
       "97    https://en.wikipedia.org/?oldid=0       113758...   \n",
       "177   https://en.wikipedia.org/?oldid=0       113758...   \n",
       "1156  https://en.wikipedia.org/?oldid=0       113758...   \n",
       "1207  https://en.wikipedia.org/?oldid=0       113758...   \n",
       "450   https://en.wikipedia.org/?oldid=0       113758...   \n",
       "1082  https://en.wikipedia.org/?oldid=0       113758...   \n",
       "1027  https://en.wikipedia.org/?oldid=0       113758...   \n",
       "709   https://en.wikipedia.org/?oldid=0       113758...   \n",
       "1183  https://en.wikipedia.org/?oldid=0       113758...   \n",
       "\n",
       "                                                 title  \\\n",
       "695                   List of sequenced animal genomes   \n",
       "226                         Neanderthal genome project   \n",
       "1210  Biological effects of radiation on the epigenome   \n",
       "1227                                              Gene   \n",
       "96                                     DeCODE genetics   \n",
       "161                    Clinical metagenomic sequencing   \n",
       "95                                     DeCODE genetics   \n",
       "449                                Rat Genome Database   \n",
       "230                               Neanderthal genetics   \n",
       "173                                          Economics   \n",
       "97                                     DeCODE genetics   \n",
       "177                                          Economics   \n",
       "1156              Vertebrate Genome Annotation Project   \n",
       "1207  Biological effects of radiation on the epigenome   \n",
       "450                                Rat Genome Database   \n",
       "1082                               Population genetics   \n",
       "1027                                    Genome editing   \n",
       "709                                          BGI Group   \n",
       "1183                                       Epigenetics   \n",
       "\n",
       "                                            heading  \\\n",
       "695                                       Bony fish   \n",
       "226                                        Findings   \n",
       "1210            The lethal-potentially lethal model   \n",
       "1227              Conflicting definitions of 'gene'   \n",
       "96         Discoveries and scientific contributions   \n",
       "161                             Laboratory workflow   \n",
       "95                               Genome of a Nation   \n",
       "449                                            Data   \n",
       "230   Interbreeding with anatomically modern humans   \n",
       "173              Definitions of economics over time   \n",
       "97                               Product innovation   \n",
       "177                          Neoclassical economics   \n",
       "1156                           Comparative analysis   \n",
       "1207                         The three lambda model   \n",
       "450                   Genome tools developed at RGD   \n",
       "1082                                  Genetic drift   \n",
       "1027                          Zinc finger nucleases   \n",
       "709                                         History   \n",
       "1183                                         Memory   \n",
       "\n",
       "                                                content  tokens  \n",
       "695   Order Anabantiformes\\nBetta splendens, Siamese...  1496.0  \n",
       "226   The researchers recovered ancient DNA of Neand...  1481.0  \n",
       "1210  This equation explores the hypothesis of a les...  1464.0  \n",
       "1227  There are lots of different ways to use the te...  1294.0  \n",
       "96    Genome research in general, and deCODE's globa...  1277.0  \n",
       "161   A typical mNGS workflow consists of the follow...  1275.0  \n",
       "95    By the time Bill Clinton and Tony Blair announ...  1268.0  \n",
       "449   RGD’s data consists of manual annotations from...  1239.0  \n",
       "230   Researchers addressed the question of possible...  1237.0  \n",
       "173   The earlier term for the discipline was 'polit...  1231.0  \n",
       "97    deCODE's scientific leadership over more than ...  1209.0  \n",
       "177   At the dawn as a social science, economics was...  1188.0  \n",
       "1156  In addition to full genomes, and unlike other ...  1182.0  \n",
       "1207  This model showed the accuracy of survival des...  1129.0  \n",
       "450   RGD develops web-based tools designed to use t...  1106.0  \n",
       "1082  Genetic drift is a change in allele frequencie...  1103.0  \n",
       "1027  As opposed to meganucleases, the concept behin...  1072.0  \n",
       "709   Wang Jian, Yu Jun, Yang Huanming and Liu Siqi ...  1071.0  \n",
       "1183  Memory formation and maintenance are due to ep...  1041.0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The number of tokens processable by gpt2 is 1024\n",
    "# I'm not sure what the downstream effects of these might be \n",
    "df[df['tokens'] > 1024].sort_values('tokens', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11752c37",
   "metadata": {},
   "outputs": [],
   "source": [
    "# It seems like all of this could be consolidated into a pretty versitile package for \n",
    "# extracting data from wikipedia on a given topic."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
